### ISSUE TITLE
Cache reduce and marginalize

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#423]
There are many instances (ex: while generating samples) when reduce or marginalize if called multiple times with the same arguments. Therefore caching the results is a good idea.

**Comment by khalibartan** [comment]
@ankurankan I would like to work on this. I have a very naive implementation, I was thinking of maintaining an additional dictionary attribute in factor class and storing the results in them. Do you have any other implementation in mind.

**Comment by ankurankan** [comment]
@khalibartan I want a more generic solution. Something like a decorator which we can put on any method that we want to cache. Something like `lru_cache` (https://docs.python.org/3/library/functools.html#functools.lru_cache)

**Comment by khalibartan** [comment]
@ankurankan I'll create my own decorator, since `lru_cache`, don't work with non hashable type of arguments and in our case arguments are mostly lists. Also I wasn't able to find `lru_cache` implementation in python2.7

**Comment by khalibartan** [comment]
@ankurankan I have implemented a decorator for `pseudo lru cache` but I wasn't able to work around with `inplace` argument.

**Comment by yashu-seth** [comment]
@khalibartan Can you tell me what was the problem you faced with the inplace argument? Were you trying to write a generic decorator for this?

**Comment by khalibartan** [comment]
@yashu-seth Yes, I was trying to write a generic decorator for this. I cannot remember what was issue with the inplace argument. here is the link to pull https://github.com/pgmpy/pgmpy/pull/598

**Comment by yashu-seth** [comment]
@khalibartan Do you still plan to work on this or should I look into this?

**Comment by khalibartan** [comment]
@ankurankan Told me to leave this issue, as it is of no use now.

**Comment by yashu-seth** [comment]
@khalibartan @ankurankan But wouldn't this save a lot of extra calls?

**Comment by khalibartan** [comment]
@yashu-seth Now I remember what was the issue with `inplace` argument . If `inplace = True` than hash value would be changed and we will not be able to use the cached value.
